openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -900,6 +907,102 @@ impl<T, B: HardwareIsolatedBacking> hv1_hypercall::ModifyVtlProtectionMask
     }
 }
 
+impl<T, B: HardwareIsolatedBacking> hv1_hypercall::TranslateVirtualAddressX64
+    for UhHypercallHandler<'_, '_, T, B>
+{
+    fn translate_virtual_address(
+        &mut self,
+        partition_id: u64,
+        vp_index: u32,
+        control_flags: hvdef::hypercall::TranslateGvaControlFlagsX64,
+        gva_page: u64,
+    ) -> HvResult<hvdef::hypercall::TranslateVirtualAddressOutput> {
+        if partition_id != hvdef::HV_PARTITION_ID_SELF {
+            return Err(HvError::AccessDenied);
+        }
+
+        if vp_index != hvdef::HV_VP_INDEX_SELF && vp_index != self.vp.vp_index().index() {
+            return Err(HvError::AccessDenied);
+        }
+
+        let target_vtl = self
+            .target_vtl_no_higher(
+                control_flags
+                    .input_vtl()
+                    .target_vtl()?
+                    .unwrap_or(self.intercepted_vtl.into()),
+            )
+            .map_err(|_| HvError::AccessDenied)?;
+
+        if self.intercepted_vtl == target_vtl {
+            return Err(HvError::AccessDenied);
+        }
+
+        let gva = gva_page * hvdef::HV_PAGE_SIZE;
+
+        let registers = self.vp.backing.translation_registers(self.vp, target_vtl);
+
+        let pat = self
+            .vp
+            .access_state(target_vtl.into())
+            .cache_control()
+            .map_err(Self::reg_access_error_to_hv_err)?
+            .msr_cr_pat;
+
+        if control_flags.tlb_flush_inhibit() {
+            self.vp
+                .set_tlb_lock(self.intercepted_vtl.into(), target_vtl);
+        }
+
+        match virt_support_x86emu::translate::translate_gva_to_gpa(
+            &self.vp.partition.gm[target_vtl], // TODO GUEST VSM: This doesn't have VTL access checks.
```

Comment:
It doesn't?

Reply:
As far as I'm aware, we haven't implemented it for hardware CVMs yet. Did I miss something?

Reply:
Oh, I guess we don't even support HvModifyVtlProtectionMask yet.

Reply:
We do support handling of HvModifyVtlProtectionMask. But to support full VTL access checks, we need to either use rmpquery (for SNP) or start tracking the VTL permissions of each page. We don't have either yet, but it will come later.

Reply:
I see. The plan must be to have a bitmap. We can't afford rmpquery calls on every guest memory access.

Reply:
My current thinking is to do the rmpquery implementation for initial bringup at least because it should be quicker to implement, and we can evaluate the performance of it. The concern with tracking vtl permissions of each page is the memory overhead, and the fact that we would be taking up extra memory even if the guest decides not to enable guest vsm. But I remember you had concerns about it being on the fast path.